### ISSUE TITLE
SStableloader fails to load a subset of BatchStatement inserted rows

### ISSUE THREAD

**Issue by joshpurvis** [scylladb/scylla-tools-java#13]
I'm running into a very strange (but reproducible) issue when trying to import via the sstablesloader found in the latest Scylla docker image. 

- It only happens to a portion of any data that was inserted via batch statements using the python-based cassandra driver (tried 2.5, 2.6, 3.3 and a few more). [maybe it's all language drivers that use that same protocol?]
- If I insert using raw CQL batch statements via cqlsh, it works. 
- If I copy the same SStables, and reload them into a clean cassandra 2.1.8 container, it works. 

Only when I try to load the SStables into Scylla does a subset of the row fail to restore... and only if they were originally inserted via a batch statement.

Here are the rough steps I follow on each permutation of my tests:

1. Clear all data directories and SStables
2. Start a single node apache cassandra 2.1.8 docker container
3. Create my keyspace and schema, via cqlsh
4. Run following script against it to insert some data
5. Run ./bin/nodetool flush, to persist from memtable to sstables.
6. Stop the cassandra container
7. Start a Scylla container (latest on docker hub), with sstable volume mounted
8. Create my keyspace and schema, via cqlsh
9. Load the SStables `./bin/sstableloader -v -d 172.17.0.2 /var/sstables/local/tree-....`
10. Check query for missing rows

I boiled my code down to this simple script, with two variations of the inserts (batch and simple):

```
from cassandra.cluster import Cluster
from cassandra.query import BatchStatement, SimpleStatement

cluster = Cluster(['localhost'])
session = cluster.connect('local')

def batch_insert(pairs):
    batch = BatchStatement()
    for root, name in pairs:
        cql = "INSERT INTO tree (root, name) VALUES (%s, %s)"
        params = (root, name)
        batch.add(SimpleStatement(cql), params)
    session.execute(batch)

def simple_insert(pairs):
    for root, name in pairs:
        cql = "INSERT INTO tree (root, name) VALUES (%s, %s)"
        params = (root, name)
        session.execute(cql, params)

root_name_pairs = [
    ('test', 'test'),
    ('test', 'test.nottest'),
    ('test.nottest', 'test.nottest'),
    ('test.nottest', 'test.nottest.basic'),
    ('test.nottest.basic', 'test.nottest.basic')
]
```

If all data is inserted directly using `simple_insert`, basically inserting one at a time, it works in every case:

```
simple_insert(root_name_pairs)
```

```
cqlsh> select * from local.tree;

 root               | name               | is_alias | is_data | target | timezone
--------------------+--------------------+----------+---------+--------+----------
               test |               test |     null |    null |   null |     null
               test |       test.nottest |     null |    null |   null |     null
       test.nottest |       test.nottest |     null |    null |   null |     null
       test.nottest | test.nottest.basic |     null |    null |   null |     null
 test.nottest.basic | test.nottest.basic |     null |    null |   null |     null
```

If data is inserted via BatchStatement, it inserts, but after flushing, exporting SStables, and sstableloading, Scylla only loads a subset of the rows:
```
batch_insert(root_name_pairs)
```

```
    cqlsh> select * from local.tree;

     root               | name               | is_alias | is_data | target | timezone
    --------------------+--------------------+----------+---------+--------+----------
                   test |               test |     null |    null |   null |     null
           test.nottest |       test.nottest |     null |    null |   null |     null
     test.nottest.basic | test.nottest.basic |     null |    null |   null |     null

    (3 rows)

```

However, I can take that same SStable and load it into Cassandra 2.1.8 and it's fine.

Here's my current schema
```
CREATE KEYSPACE IF NOT EXISTS local
WITH REPLICATION = {
    'class' : 'SimpleStrategy',
    'replication_factor': 1
};

use local;

CREATE TABLE IF NOT EXISTS tree (
    root text,
    name text,
    target text,
    timezone text,
    is_data boolean,
    is_alias boolean,
    PRIMARY KEY ((root), name)
);
```

Thanks!

**Comment by penberg** [comment]
@avikivity @elcallio

**Comment by joshpurvis** [comment]
If it helps, I've recently narrowed this issue down a bit further. Ignoring the drivers, and just focusing on pure `BATCH` statements in `CQL` -- It seems if data is loaded without a `USING TIMESTAMP <x>`, the issue manifests itself:

```
BEGIN BATCH
  INSERT INTO local.tree (root,name) values ('test','test');
  INSERT INTO local.tree (root,name) values ('test','test.nottest');
  INSERT INTO local.tree (root,name) values ('test.nottest','test.nottest');
  INSERT INTO local.tree (root,name) values ('test.nottest','test.nottest.basic');
  INSERT INTO local.tree (root,name) values ('test.nottest.basic','test.nottest.basic');
APPLY BATCH;
```

^ This results in only 3 rows when its sstables are loaded into Scylla.

However, if you provide `USING TIMESTAMP <x>` on each individual query, like this:

```
BEGIN BATCH
  INSERT INTO local.tree (root,name) values ('test','test') USING TIMESTAMP 1476214949653175;
  INSERT INTO local.tree (root,name) values ('test','test.nottest') USING TIMESTAMP 1476214949654351;
  INSERT INTO local.tree (root,name) values ('test.nottest','test.nottest') USING TIMESTAMP 1476214949647359;
  INSERT INTO local.tree (root,name) values ('test.nottest','test.nottest.basic') USING TIMESTAMP 1476214949651980;
  INSERT INTO local.tree (root,name) values ('test.nottest.basic','test.nottest.basic') USING TIMESTAMP 1476214949650750;
APPLY BATCH;
```

This time, all 5 rows are successfully loaded into Scylla. 

With this, it makes me wonder if there's something overlooked in the SSTableToCQL conversion when queries within a batch don't include the timestamp?

**Comment by elcallio** [comment]
The cql row checker in SSTableToCQL was never updated, leading to the broken clustering above. Of course, separating timestamps (as you did) will work around this by forcing the cells ranges to be processed individually, which is why the second case above worked. 
I've sent a patch.

**Comment by joshpurvis** [comment]
Thanks @elcallio ! 

Out of curiosity, could you show me where I can see the patch/commit?

**Comment by joshpurvis** [comment]
Just an update if anyone else runs into this, the above patch fixed it. 

I ran my tests against the `sstableloader` found in the nightly docker image (scylladb/scylla-nightly) and it imports correctly now.

Thanks guys!

**Comment by penberg** [comment]
@jasobrown Thanks for reporting back!

**Comment by penberg** [comment]
The fix is scheduled for 1.3.3 and 1.4.0.
